src/fasta.rs

```diff
@@ -0,0 +1,210 @@
+                    title = Some(c.as_str().trim().to_string());
+                } else {
+                    match title {
+                        None => {
+                            // There is no title, skip this content!
+                        }
+                        Some(_) => {
+                            lines.push(line.replace(' ', "").replace('\r', ""));
+                        }
+                    }
+                }
+            }
+        }
+
+        if let Some(t) = title {
+            result.push((t, lines.join("")));
+        }
+
+        Ok(result)
+    }
+}
+
+pub struct DNAFastaParser {}
+
+impl DNAFastaParser {
+    pub fn parse<R: BufRead>(handle: &mut R) -> Result<Vec<(String, DnaSequence)>, anyhow::Error> {
+        SimpleFastaParser::parse(handle)?
+            .into_iter()
+            .map(|(title, sequence)| -> Result<(String, DnaSequence), _> {
+                let dna = DnaSequence::from_str(&*sequence)?;
+                Ok((title, dna))
+            })
+            .collect()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::fasta::{DNAFastaParser, SimpleFastaParser};
+    use crate::DnaSequence;
+    use std::io::BufReader;
+    use std::str::FromStr;
+
+    #[test]
+    fn test_empty_fasta() {
+        let r = SimpleFastaParser::parse(&mut BufReader::new("".as_bytes())).unwrap();
+        assert_eq!(r.len(), 0);
+    }
+
+    #[test]
+    fn test_empty_fasta_with_newlines() {
+        let r = SimpleFastaParser::parse(&mut BufReader::new("\n\n".as_bytes())).unwrap();
+        assert_eq!(r.len(), 0);
+    }
+
+    #[test]
+    fn test_empty_fasta_with_many_newlines() {
+        let r = SimpleFastaParser::parse(&mut BufReader::new(
+            "  \n\n \n  \r\r \r\n  \r \n".as_bytes(),
+        ))
+        .unwrap();
+        assert_eq!(r.len(), 0);
+    }
+
+    #[test]
+    fn test_fasta_with_no_content() {
+        let r = SimpleFastaParser::parse(&mut BufReader::new(">Virus\n".as_bytes())).unwrap();
+        assert_eq!(r.len(), 1);
+        assert_eq!(r[0], ("Virus".to_string(), "".to_string()));
+    }
+
+    #[test]
+    fn test_fasta_with_empty_content() {
+        let r = SimpleFastaParser::parse(&mut BufReader::new(">Virus\n\n".as_bytes())).unwrap();
+        assert_eq!(r.len(), 1);
+        assert_eq!(r[0], ("Virus".to_string(), "".to_string()));
+    }
+
+    #[test]
+    fn test_fasta_with_stuff_no_header() {
+        let r = SimpleFastaParser::parse(&mut BufReader::new(
+            "// this is a file comment\n@author is foo\n\n".as_bytes(),
+        ))
+        .unwrap();
+        assert_eq!(r.len(), 0);
+    }
+
+    #[test]
+    fn test_fasta_with_stuff_before_header() {
+        let r = SimpleFastaParser::parse(&mut BufReader::new(
+            "// this is a file comment\n@author is foo\n\n>Virus\n\n".as_bytes(),
+        ))
+        .unwrap();
+        assert_eq!(r.len(), 1);
+        assert_eq!(r.len(), 1);
+        assert_eq!(r[0], ("Virus".to_string(), "".to_string()));
```

Comment:
```suggestion
        assert_eq!(r, vec![("Virus".to_string(), "".to_string())]);
```

Reply:
I'd prefer that more declarative style for the tests, though the tests are otherwise great